contrib/scalacheck/shared/src/main/scala/eu/timepit/refined/scalacheck/numeric.scala

```diff
@@ -3,92 +3,165 @@ package scalacheck
+  ): Arbitrary[F[T, Greater[N]]] =
+    rangeOpenClosedArbitrary(nt.fromInt(tn()), bounded.maxValue)
+
+  implicit def greaterEqualArbitraryWit[F[_, _]: RefType, T: Numeric: Choose, N <: T](
+    implicit
+    bounded: Bounded[T],
+    wn: Witness.Aux[N]
+  ): Arbitrary[F[T, GreaterEqual[N]]] =
+    rangeClosedArbitrary(wn.value, bounded.maxValue)
 
-  implicit def greaterEqualArbitrary[F[_, _], T, N <: T](
+  implicit def greaterEqualArbitraryNat[F[_, _]: RefType, T: Choose, N <: Nat](
     implicit
-    rt: RefType[F],
-    wn: Witness.Aux[N],
+    bounded: Bounded[T],
     nt: Numeric[T],
-    c: Gen.Choose[T]
-  ): Arbitrary[F[T, GreaterEqual[N]]] = {
-    val gen = Gen.chooseNum(wn.value, nt.fromInt(Int.MaxValue))
-    arbitraryRefType(gen)
-  }
+    tn: ToInt[N]
+  ): Arbitrary[F[T, GreaterEqual[N]]] =
+    rangeClosedArbitrary(nt.fromInt(tn()), bounded.maxValue)
+
+  ///
 
-  implicit def intervalOpenArbitrary[F[_, _], T, L <: T, H <: T](
+  implicit def intervalOpenArbitrary[F[_, _]: RefType, T: Numeric: Choose, L <: T, H <: T](
     implicit
-    rt: RefType[F],
     wl: Witness.Aux[L],
-    wh: Witness.Aux[H],
-    nt: Numeric[T],
-    c: Gen.Choose[T]
+    wh: Witness.Aux[H]
   ): Arbitrary[F[T, Interval.Open[L, H]]] =
-    arbitraryRefType(Gen.chooseNum(wl.value, wh.value).filter(t => nt.gt(t, wl.value) && nt.lt(t, wh.value)))
+    rangeOpenArbitrary(wl.value, wh.value)
 
-  implicit def intervalOpenClosedArbitrary[F[_, _], T, L <: T, H <: T](
+  implicit def intervalOpenClosedArbitrary[F[_, _]: RefType, T: Numeric: Choose, L <: T, H <: T](
     implicit
-    rt: RefType[F],
     wl: Witness.Aux[L],
-    wh: Witness.Aux[H],
-    nt: Numeric[T],
-    c: Gen.Choose[T]
+    wh: Witness.Aux[H]
   ): Arbitrary[F[T, Interval.OpenClosed[L, H]]] =
-    arbitraryRefType(Gen.chooseNum(wl.value, wh.value).filter(t => nt.gt(t, wl.value)))
+    rangeOpenClosedArbitrary(wl.value, wh.value)
 
-  implicit def intervalClosedOpenArbitrary[F[_, _], T, L <: T, H <: T](
+  implicit def intervalClosedOpenArbitrary[F[_, _]: RefType, T: Numeric: Choose, L <: T, H <: T](
     implicit
-    rt: RefType[F],
     wl: Witness.Aux[L],
-    wh: Witness.Aux[H],
-    nt: Numeric[T],
-    c: Gen.Choose[T]
+    wh: Witness.Aux[H]
   ): Arbitrary[F[T, Interval.ClosedOpen[L, H]]] =
-    arbitraryRefType(Gen.chooseNum(wl.value, wh.value).filter(t => nt.lt(t, wh.value)))
+    rangeClosedOpenArbitrary(wl.value, wh.value)
 
-  implicit def intervalClosedArbitrary[F[_, _], T, L <: T, H <: T](
+  implicit def intervalClosedArbitrary[F[_, _]: RefType, T: Numeric: Choose, L <: T, H <: T](
     implicit
-    rt: RefType[F],
     wl: Witness.Aux[L],
-    wh: Witness.Aux[H],
-    nt: Numeric[T],
-    c: Gen.Choose[T]
+    wh: Witness.Aux[H]
   ): Arbitrary[F[T, Interval.Closed[L, H]]] =
-    arbitraryRefType(Gen.chooseNum(wl.value, wh.value))
+    rangeClosedArbitrary(wl.value, wh.value)
+
+  ///
+
+  private def rangeOpenArbitrary[F[_, _]: RefType, T: Choose, P](min: T, max: T)(
```

Comment:
We can also make them public, I've no problem with that.

I think I now better understand the purpose of `greaterArbitraryRange`. That is something you would use explicitly when writing a property, right? So you would write `property("xyz") = forAll(greaterArbitraryRange(Natural(0), Natural(10)).arbitrary) { n => ??? }`. I think it should be then a `Gen` instead of an `Arbitrary`.

Maybe we can also have a variant of `greaterArbitraryRange` that supports all numeric predicates:
```scala
def closedRangeWithBounds[F[_, _], T: Numeric: Choose, P](min: F[T, P], max: F[T, P])(
  implicit rt: RefType[F]
): Gen[F[T, P]] =
  Gen.choose(rt.unwrap(min), rt.unwrap(max)).map(rt.unsafeWrap)
```
This _should_ work with predicates that are indexed by literal singleton types and `Nat`s. Usage would then look like this
```scala
forAll(closedRangeWithBounds(Natural(0), Natural(10))) { n =>
  ???
}
```